Lib/fontTools/qu2cu/qu2cu.py

```diff
@@ -0,0 +1,327 @@
+    Returns:
+        bool: True if the cubic Bezier ``p`` entirely lies within a distance
+        ``tolerance`` of the origin, False otherwise.
+    """
+    # First check p2 then p1, as p2 has higher error early on.
+    if abs(p2) <= tolerance and abs(p1) <= tolerance:
+        return True
+
+    # Split.
+    mid = (p0 + 3 * (p1 + p2) + p3) * 0.125
+    if abs(mid) > tolerance:
+        return False
+    deriv3 = (p3 + p2 - p1 - p0) * 0.125
+    return cubic_farthest_fit_inside(
+        p0, (p0 + p1) * 0.5, mid - deriv3, mid, tolerance
+    ) and cubic_farthest_fit_inside(mid, mid + deriv3, (p2 + p3) * 0.5, p3, tolerance)
+
+
+@cython.locals(_1_3=cython.double, _2_3=cython.double)
+@cython.locals(
+    p0=cython.complex,
+    p1=cython.complex,
+    p2=cython.complex,
+    p1_2_3=cython.complex,
+)
+def elevate_quadratic(p0, p1, p2, _1_3=1 / 3, _2_3=2 / 3):
+    """Given a quadratic bezier curve, return its degree-elevated cubic."""
+
+    # https://pomax.github.io/bezierinfo/#reordering
+    p1_2_3 = p1 * _2_3
+    return (
+        p0,
+        (p0 * _1_3 + p1_2_3),
+        (p2 * _1_3 + p1_2_3),
+        p2,
+    )
+
+
+@cython.locals(
+    n=cython.int,
+    k=cython.int,
+    prod_ratio=cython.double,
+    sum_ratio=cython.double,
+    ratio=cython.double,
+    p0=cython.complex,
+    p1=cython.complex,
+    p2=cython.complex,
+    p3=cython.complex,
+)
+def merge_curves(curves):
+    """Give a cubic-Bezier spline, reconstruct one cubic-Bezier
+    that has the same endpoints and tangents and approxmates
+    the spline."""
+
+    # Reconstruct the t values of the cut segments
+    n = len(curves)
+    prod_ratio = 1.0
+    sum_ratio = 1.0
+    ts = [1]
+    for k in range(1, n):
+        ck = curves[k]
+        c_before = curves[k - 1]
+
+        # |t_(k+1) - t_k| / |t_k - t_(k - 1)| = ratio
+        assert ck[0] == c_before[3]
+        ratio = abs(ck[1] - ck[0]) / abs(c_before[3] - c_before[2])
+
+        prod_ratio *= ratio
+        sum_ratio += prod_ratio
+        ts.append(sum_ratio)
+
+    # (t(n) - t(n - 1)) / (t_(1) - t(0)) = prod_ratio
+
+    ts = [t / sum_ratio for t in ts[:-1]]
+
+    p0 = curves[0][0]
+    p1 = curves[0][1]
+    p2 = curves[n - 1][2]
+    p3 = curves[n - 1][3]
+
+    # Build the curve by scaling the control-points.
+    p1 = p0 + (p1 - p0) / (ts[0] if ts else 1)
+    p2 = p3 + (p2 - p3) / ((1 - ts[-1]) if ts else 1)
+
+    curve = (p0, p1, p2, p3)
+
+    return curve, ts
+
+
+def add_implicit_on_curves(p):
+    q = list(p)
+    count = 0
+    num_offcurves = len(p) - 2
+    for i in range(1, num_offcurves):
+        off1 = p[i]
+        off2 = p[i + 1]
+        on = off1 + (off2 - off1) * 0.5
+        q.insert(i + 1 + count, on)
+        count += 1
+    return q
+
+
+def quadratics_to_curves(pp, tolerance=0.5, all_cubic=False):
+    is_complex = type(pp[0][0]) is complex
+    if not is_complex:
+        pp = [[complex(x, y) for (x, y) in p] for p in pp]
+
+    q = [pp[0][0]]
+    cost = 0
+    costs = [0]
```

Comment:
it's unclear what these `costs` are referring to or how they are used, can you please add a comment at least to help the reader?